aequilibrae/paths/graph.py

```diff
@@ -103,25 +109,206 @@ def prepare_graph(self, centroids: np.ndarray) -> None:
-                        if centroids.shape[0] != np.unique(centroids).shape[0]:
-                            raise ValueError("Centroid IDs are not unique")
-                self.centroids = np.array(list(centroids), np.uint32)
-            else:
-                raise ValueError("Centroids need to be an array of integers 64 bits")
-        else:
+
+        if centroids is None or not isinstance(centroids, np.ndarray):
+            raise ValueError("Centroids need to be a NumPy array of integers 64 bits")
+        if not np.issubdtype(centroids.dtype, np.integer):
             raise ValueError("Centroids need to be a NumPy array of integers 64 bits")
+        if centroids.shape[0] == 0:
+            raise ValueError("You need at least one centroid")
+        if centroids.min() <= 0:
+            raise ValueError("Centroid IDs need to be positive")
+        if centroids.shape[0] != np.unique(centroids).shape[0]:
+            raise ValueError("Centroid IDs are not unique")
+        self.centroids = np.array(centroids, np.uint32)
+
+        properties = self.__build_directed_graph(self.network, centroids)
+        self.all_nodes, self.num_nodes, self.nodes_to_indices, self.fs, self.graph = properties
+        self.num_links = self.graph.shape[0]
         self.__build_derived_properties()
 
-        all_titles = list(self.network.columns)
-
-        not_pos = self.network.loc[self.network.direction != 1, :]
-        not_negs = self.network.loc[self.network.direction != -1, :]
+        self.__build_compressed_graph()
+        self.compact_num_links = self.compact_graph.shape[0]
+
+    def __build_compressed_graph(self):
+        # Build link index
+        link_idx = np.empty(self.network.link_id.max() + 1).astype(np.int)
+        link_idx[self.network.link_id] = np.arange(self.network.shape[0])
+
+        nodes = np.hstack([self.network.a_node.values, self.network.b_node.values])
+        links = np.hstack([self.network.link_id.values, self.network.link_id.values])
+
+        idx = np.argsort(nodes)
+        all_nodes = nodes[idx]
+        all_links = links[idx]
+        links_index = np.empty(all_nodes.max() + 1, np.int64)
+        links_index.fill(-1)
+        nlist = np.arange(all_nodes.max())
+
+        y, x, _ = np.intersect1d(all_nodes, nlist, assume_unique=False, return_indices=True)
+        links_index[y] = x[:]
+        links_index[-1] = all_links.shape[0]
+        for i in range(all_nodes.max(), 1, -1):
+            links_index[i - 1] = links_index[i] if links_index[i - 1] == -1 else links_index[i - 1]
+
+        nodes = np.hstack([self.network.a_node.values, self.network.b_node.values])
+        counts = np.bincount(nodes)
+        counts[self.centroids] = 999
+
+        truth = (counts == 2).astype(np.int)
+        link_edge = truth[self.network.a_node.values] + truth[self.network.b_node.values]
+        link_edge = self.network.link_id.values[link_edge == 1]
+
+        simplified_links = np.empty(self.network.link_id.max() + 1)
```

Comment:
maybe `np.repeat(-1, self.network.link_id.max()+1)` instead of these l.162-164?

Reply:
Good one. I am incorporating it